docs/3 Command line options.MD
# Command line options

## Commands

Commands are issued as the first parameter on the command line and set
the task to be run by the program.

  - `makedb`  
      
    Create a DIAMOND formatted reference database from a FASTA input
    file.

  - `blastp`  
      
    Align protein query sequences against a protein reference database.

  - `blastx`  
      
    Align translated DNA query sequences against a protein reference
    database.

  - `view`  
      
    Generate formatted output from DAA files.

  - `version`  
      
    Print version information.

  - `dbinfo`  
      
    Print information about a database file.

  - `help`  
      
    Print help message.
    
  - `test`
  
    Run a series of test cases and verify the output against reference
    hashes. This command will exit with code `0` if all
    tests have passed and `1` otherwise. Running this command requires write
    access to the current working directory.

## Makedb options

  - `--in <file>`  
      
    Path to the input protein reference database file in FASTA format
    (may be gzip compressed). If this parameter is omitted, the input
    will be read from `stdin`.

  - `--db/-d <file>`  
      
    Path to the output DIAMOND database file.

  - `--taxonmap <file>`  
      
    Path to mapping file that maps NCBI protein accession numbers to
    taxon ids (gzip compressed). This parameter is optional and needs to
    be supplied in order to provide taxonomy features. The file can be
    downloaded from NCBI: <ftp://ftp.ncbi.nlm.nih.gov/pub/taxonomy/accession2taxid/prot.accession2taxid.gz>.
    
    A custom file following the same format may be supplied here. Note
    that the first line of this file is assumed to contain headings and
    will be ignored.

  - `--taxonnodes <file>`  
      
    Path to the `nodes.dmp` file from the NCBI taxonomy. This parameter
    is optional and needs to be supplied in order to provide taxonomy
    features. The file is contained within this archive downloadable at
    NCBI: <ftp://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdmp.zip>.
    
  - `--taxonnames <file>`
    
    Path to the `names.dmp` file from the NCBI taxonomy. This parameter
    is optional and needs to be supplied in order to provide taxonomy
    features. The file is contained within this archive downloadable at
    NCBI: <ftp://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdmp.zip>.

## General options

  - `--threads/-p #`  
      
    Number of CPU threads. By default, the program will auto-detect and
    use all available virtual cores on the machine.
    
  - `--quiet`
  
    Disable all terminal output.
    
  - `--verbose/-v`
  
    Enable more verbose terminal output.
    
  - `--log`
  
    Enable even more verbose terminal output, which is also written to
    a file named `diamond.log` is the current working directory.

## Input options

  - `--db/-d <file>`  
      
    Path to the DIAMOND database file.

  - `--query/-q <file>`  
      
    Path to the query input file in FASTA or FASTQ format (may be gzip
    compressed). If this parameter is omitted, the input will be read
    from `stdin`.

  - `--taxonlist <list>`  
      
    Comma-separated list of NCBI taxonomic IDs to filter the database
    by. Any taxonomic rank can be used, and only reference sequences
    matching one of the specified taxon ids will be searched against.
    Using this option requires setting the `--taxonmap` and `--taxonnodes`
    parameters for `makedb`.

  - `--query-gencode #`  
      
    Genetic code used for translation of query in BLASTX mode. A list of
    possible values can be found at the [NCBI
    website](https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi).
    By default, the Standard Code is used. Note: changing the genetic
    code is currently not fully supported for the DAA format.

  - `--strand {both, plus, minus}`  
      
    Set strand of query to align for translated searches. By default
    both strands are searched.

  - `--min-orf/-l #`  
      
    Ignore translated sequences that do not contain an open reading
    frame of at least this length. By default this feature is disabled
    for sequences of length below 30, set to 20 for sequences of length
    below 100, and set to 40 otherwise. Setting this option to `1` will
    disable this feature.

## Alignment options

  - `--mid-sensitive`
  
    Enable the mid-sensitive mode which is between the default (fast) mode
    and the sensitive mode in sensitivity. *Option supported since v2.0.3*
  
    Without using any sensitivity option, the default (fast) mode will run which
    is designed for finding hits of >70% identity and short read alignment.

  - `--sensitive`
  
    Enable the sensitive mode designed for full sensitivity for hits >40% identity.
    
  - `--more-sensitive`  
      
    This mode is slightly more sensitive than the `--sensitive` mode.
    
  - `--very-sensitive`
  
    Enable the very-sensitive mode designed for best sensitivity including the
    twilight zone range of <40% identity.
    
    **Available since version 2.0.0.**
    
  - `--ultra-sensitive`
    
    Enable the ultra-sensitive mode which is yet more sensitive than the
    `--very-sensitive` mode.
    
    **Available since version 2.0.0.**

  - `--frameshift/-F #`  
      
    Penalty for frameshifts in DNA-vs-protein alignments. Values around
    15 are reasonable for this parameter. Enabling this feature will
    have the aligner tolerate missing bases in DNA sequences and is most
    recommended for long, error-prone sequences like MinION reads.  
    In the pairwise output format, frameshifts will be indicated by `\`
    and `/` for a shift by +1 and -1 nucleotide in the direction of
    translation respectively.
    Note that this feature is disabled by default.

  - `--gapopen #`  
      
    Gap open penalty.

  - `--gapextend #`  
      
    Gap extension penalty.

  - `--matrix <matrix name>`  
      
    Scoring matrix. The following matrices are supported, with the
    default being
    BLOSUM62.  
      
    
    | Matrix   | Supported values for (gap open)/(gap extend) | Default gap penalties |
    | :------- | :------------------------------------------- | :-------------------- |
    | BLOSUM45 | (10-13)/3; (12-16)/2; (16-19)/1              | 14/2                  |
    | BLOSUM50 | (9-13)/3; (12-16)/2; (15-19)/1               | 13/2                  |
    | BLOSUM62 | (6-11)/2; (9-13)/1                           | 11/1                  |
    | BLOSUM80 | (6-9)/2; 13/2; 25/2; (9-11)/1                | 10/1                  |
    | BLOSUM90 | (6-9)/2; (9-11)/1                            | 10/1                  |
    | PAM250   | (11-15)/3; (13-17)/2; (17-21)/1              | 14/2                  |
    | PAM70    | (6-8)/2; (9-11)/1                            | 10/1                  |
    | PAM30    | (5-7)/2; (8-10)/1                            | 9/1                   |
    
  - `--masking (0,1)`
  
    DIAMOND by default applies the tantan repeat masking algorithm to the
    query and target sequences as described in [1]. This masking procedure
    increases the specificity of alignments and serves to filter out
    spurious hits. If this is not desired, repeat masking can be disabled
    using `--masking 0`.
  
  - `--comp-based-stats (0,1)`  
      
    Compositional bias correction of alignment scores. 0 means no score
    correction, 1 means compositional bias correction as described in 
    [2]. Compositionally biased sequences often cause false
    positive matches, which are effectively filtered by this algorithm
    in a way similar to the composition based statistics used by BLAST .

  - `--algo (0,1)`  
      
    Algorithm for seed search. 0 means double-indexed and 1 means
    query-indexed. The double-indexed algorithm is the program’s main
    algorithm, but it is inefficient for very small query files, where
    the query-indexed algorithm should be used instead.  
    By default, the program will automatically choose one of the
    algorithms based on the size of the query and database files. The
    algorithm used will be displayed at program startup.  
    Note that while the two algorithms are configured to provide roughly
    the same sensitivity for the respective modes, results will not be
    exactly identical to each other.

## Output options

  - `--out/-o <file>`  
      
    Path to the output file. If this parameter is omitted, the results
    will be written to the standard output and all other program output
    will be suppressed.

  - `--outfmt/-f #`  
      
    Format of the output file. The following values are accepted:
    
      - `0`  
        BLAST pairwise format.
    
      - `5`  
        BLAST XML format.
    
      - `6`  
        BLAST tabular format (default). This format can be customized,
        the `6` may be followed by a space-separated list of the
        following keywords, each specifying a field of the output.
        
          - `qseqid`
            Query Seq - id
        
          - `qlen`
            Query sequence length
        
          - `sseqid`
            Subject Seq - id
        
          - `sallseqid`
            All subject Seq - id(s), separated by a ’;’
        
          - `slen`
            Subject sequence length
        
          - `qstart`
            Start of alignment in query*
        
          - `qend`
            End of alignment in query*
        
          - `sstart`
            Start of alignment in subject*
        
          - `send`
            End of alignment in subject*
        
          - `qseq`
            Aligned part of query sequence*
            
          - `full_qseq`
            Full query sequence
        
          - `sseq`
            Aligned part of subject sequence*
        
          - `full_sseq`
            Full subject sequence
        
          - `evalue`
            Expect value
        
          - `bitscore`
            Bit score
        
          - `score`
            Raw score
        
          - `length`
            Alignment length*
        
          - `pident`
            Percentage of identical matches*
        
          - `nident`
            Number of identical matches*
        
          - `mismatch`
            Number of mismatches*
        
          - `positive`
            Number of positive - scoring matches*
        
          - `gapopen`
            Number of gap openings*
        
          - `gaps`
            Total number of gaps*
        
          - `ppos`
            Percentage of positive - scoring matches*
        
          - `qframe`
            Query frame
        
          - `btop`
            Blast traceback operations(BTOP)*
            
          - `cigar`
            CIGAR string*
        
          - `staxids`
            Unique Subject Taxonomy ID(s), separated by a ’;’ (in
            numerical order). This field requires setting the
            `--taxonmap` parameter for `makedb`.
            
          - `sscinames`
            Unique Subject Scientific Name(s), separated by a ';'. This
            field requires setting the `--taxonmap` and `--taxonnames`
            parameters for `makedb`.
          
          - `sskingdoms`
            Unique Subject Super Kingdom(s), separated by a ';'. This
            field requires setting the `--taxonmap`, `--taxonnodes`
            and `--taxonnames` parameters for `makedb`.
            
          - `skingdoms`
            Unique Subject Kingdom(s), separated by a ';'. This
            field requires setting the `--taxonmap`, `--taxonnodes`
            and `--taxonnames` parameters for `makedb`.
            
          - `sphylums`
            Unique Subject Phylums(s), separated by a ';'. This
            field requires setting the `--taxonmap`, `--taxonnodes`
            and `--taxonnames` parameters for `makedb`.
        
          - `stitle`
            Subject Title
        
          - `salltitles`
            All Subject Title(s), separated by a ’\<\>’
        
          - `qcovhsp`
            Query Coverage Per HSP*
            
          - `scovhsp`
            Subject Coverage Per HSP*
        
          - `qtitle`
            Query title
            
          - `qqual`
            Query quality values for the aligned part of the query*
            
          - `full_qqual`
            Query quality values
            
          - `qstrand`
            Query strand
        
        By default, there are 12 preconfigured fields: `qseqid sseqid
        pident length mismatch gapopen qstart qend sstart send evalue
        bitscore`.
        
        *These fields require alignment traceback. If none of these fields
        are selected, traceback computation will be disabled, which improves
        performance and reduces use of temporary disk space.
    
      - `100`  
        DIAMOND alignment archive (DAA). The DAA format is a proprietary
        binary format that can subsequently be used to generate other
        output formats using the `view` command. It is also supported by
        MEGAN and allows a quick import of results. *Note that this format
        cannot be written to a stream output like stdout.*
    
      - `101`  
        SAM format.
    
      - `102`  
        Taxonomic classification. This format will not print alignments
        but only a taxonomic classification for each query using the LCA
        algorithm. The output lines consist of 3 tab-delimited fields:
        
        1.  Query ID
        
        2.  NCBI taxonomy ID (0 if unclassified)
        
        3.  E-value of the best alignment with a known taxonomic ID
            found for the query (0 if unclassified)
        
        The score range for the LCA algorithm is set by the `--top`
        parameter. The default value is 10 which means that all
        alignments whose score is at most 10% lower than the best score
        are considered for the LCA computation.
        
        Using this format requires setting the `--taxonmap` and
        `--taxonnodes` parameters for `makedb`.
    
      - `103`  
        [PAF format](https://github.com/lh3/miniasm/blob/master/PAF.md).
        The custom fields in the format are AS (bit score), ZR (raw
        score) and ZE (e-value).

  - `--salltitles`  
      
    Include full length subject titles into the DAA format. By default,
    DAA files contain only the shortened sequence id (up to the first
    blank character).

  - `--sallseqid`  
      
    Include all subject ids into the DAA file. By default only the first
    id of each subject is included. As the subject ids are much shorter
    than the full titles this option will save space compared to the
    `--salltitles` option.

  - `--compress (0,1)`  
      
    Enable compression of the output file. 0 (default) means no
    compression, 1 means gzip compression.

  - `--max-target-seqs/-k #`  
      
    The maximum number of target sequences per query to report
    alignments for (default=25). Setting this to `-k0` will report all
    targets for which alignments were found.
    
    *Note that this parameter does not only affect the reporting, but
    also the algorithm as it is taken into account for heuristics that
    eliminate hits prior to full gapped extension.*

  - `--top #`  
      
    Report alignments within the given percentage range of the top
    alignment score for a query (overrides `--max-target-seqs` option).
    For example, setting `--top 10` will report all alignments whose
    score is at most 10% lower than the best alignment score for a
    query.
    
    *Note that this parameter does not only affect the reporting, but
    also the algorithm as it is taken into account for heuristics that
    eliminate hits prior to full gapped extension.*
        
  - `--max-hsps #`  
      
    The maximum number of HSPs (High-Scoring Segment Pairs) per target
    sequence to report for each query. The default policy is to report
    only the highest-scoring HSP for each target, while disregarding
    alternative, lower-scoring HSPs that are contained in the same target.
    **This is not to be confused with the `--max-target-seqs` option.**
    
    This parameter can be increased to report an alternative HSP if 
    its query and subject ranges are not enveloped by a higher scoring
    HSP and if it meets the e-value threshold. Setting this option to
    `--max-hsps 0` will report all alternative HSPs.

  - `--range-culling`  

    Enable hit culling with respect to the query range. This feature is
    designed for long query DNA sequences that may span several genes.
    In these cases, reporting the overall top N hits can cause hits to
    a lower-scoring gene to be superseded by a higher-scoring gene.
    Using this option, hit culling will be performed locally with respect
    to a hit's query range, thus reporting the locally top N hits while
    allowing more hits that span a different region of the query.

    Using this feature along with `-k 25` (default), a hit will only be
    deleted if at least 50% of its query range is spanned by at least 25
    higher or equal scoring hits.
    
    Using this feature along with `--top 10`, a hit will only be deleted
    if its score is more than 10% lower than that of a higher scoring
    hit over at least 50% of its query range.
    
    The overlap percentage is configurable using `--range-cover`. Note that
    this feature is currently only available in frameshift alignment mode.

  - `--evalue/-e #`  
      
    Maximum expected value to report an alignment (default=0.001).

  - `--min-score #`  
      
    Minimum bit score to report an alignment. Setting this option will
    override the `--evalue` parameter.

  - `--id #`  
      
    Report only alignments above the given percentage of sequence
    identity.
    
    *Note that using this option reduces performance.*

  - `--query-cover #`  
      
    Report only alignments above the given percentage of query cover.
    
    *Note that using this option reduces performance.*

  - `--subject-cover #`  
      
    Report only alignments above the given percentage of subject cover.
    
    *Note that using this option reduces performance.*

  - `--unal (0,1)`  
      
    Report unaligned queries (0=no, 1=yes). By default, unaligned
    queries are reported for the BLAST pairwise, BLAST XML and SAM
    format.

  - `--no-self-hits`  
      
    Suppress reporting of identical self-hits between sequences.
    The FASTA sequence identifier as well as the sequence of
    query and target need to be identical for a hit to be deleted.

## Memory & performance options

  - `--block-size/-b #`  
      
    Block size in billions of sequence letters to be processed at a
    time. This is the main parameter for controlling the program’s
    memory and disk space usage. Bigger numbers will increase the use
    of memory and temporary disk space, but also improve performance.
    The program can be expected to use roughly six times this number
    of memory (in GB).
    
    The default value is `-b2.0`. The parameter can be decreased for
    reducing memory use, as well as increased for better performance
    (values of >20 are not recommended).
    
    The very-sensitive and ultra-sensitive modes use `-b0.4` as a
    default. Note that these two modes benefit only slightly
    from increasing this parameter.
    
    Note that this parameter affects the algorithm and results will not
    be completely identical for different values of the block size.

  - `--tmpdir/-t <directory>`  
      
    Directory to be used for temporary storage. This is set to the
    output directory by default. The amount of disk space that will be
    used depends on the program’s settings and the input data. As a general
    rule it should be ensured that 100 GB of disk space are available here.
    
    If the program is being run in a cluster environment, and disk space is
    mounted over a network based file system, it is recommended to set this
    parameter to a fast local disk or to `/dev/shm` to avoid any I/O
    bottlenecks.

  - `--index-chunks/-c #`  
      
    The number of chunks for processing the seed index. This
    option can be additionally used to tune the performance. The default
    value is `-c4`, while setting this parameter to `-c1` instead will
    improve the performance at the cost of increased memory use. Note
    that the very-sensitive and ultra-sensitive modes use `-c1` by
    default.

## View options

  - `--daa/-a <file>`  
      
    Path to input file in DAA format.

  - `--out/-o <file>`  
      
    Path to output file. If this parameter is omitted, the results will
    be written to the standard output and all other program output will
    be suppressed.

These aligner parameters apply to the view command as well and work in
the same way: `--outfmt`, `--compress`, `--max-target-seqs`, `--top`. Note
that taxonomy features are currently not available for the DAA format.

## Advanced options

  - `--freq-sd #`  
      
    During the seed search, seeds with very high frequency in the
    queries or the database are ignored. This option sets the number of
    standard deviations above the average frequency for ignoring a seed.
    The default values are 50 in default mode, 20 in sensitive mode,
    200 in more-sensitive mode, 15 in very-sensitive mode and 20 in
    ultra-sensitive mode.
    

src/ChangeLog
[2.0.6]
- Changed the computation of expected values to use the method described in Park, Y.,
  Sheetlin, S., Ma, N. et al. New finite-size correction for local alignment score
  distributions. *BMC Res Notes* **5**, 286 (2012).
- Enabled the use of a custom scoring matrix without having to specify the statistical
  parameters (option `--custom-matrix`).
- Fixed a bug that could cause incorrect alignment coordinates, gaps counts and sequence
  identities being reported by `diamond view`.
- Targets are sorted by bit score instead of e-value when the `--top` parameter is used
  in the alignment output.
- Disabled support of custom scoring matrices for the DAA format.
- Fixed a bug that caused the use of a custom scoring matrix not to function correctly.
- Fixed an issue that caused the portable binary not to function on systems that do not
  support AVX.

[2.0.5]
- Fixed an issue that could cause high memory use in frameshift alignment mode.

[2.0.4]
- Fixed a bug that could cause the `--max-target-seqs/-k`, `--ext-chunk-size` and 
  `--file-buffer-size` options not to function correctly on macOS.

[2.0.3]
- Added a new sensitivity mode that is between the default mode and the sensitive mode
  in sensitivity (option `--mid-sensitive`).
- Added counters for total number of reference blocks, shapes and index chunks to
  the status messages.
- Fixed a bug (persisting since v2.0.2) that could cause secondary HSPs within one
  target not to be reported if the `--max-hsps` option was used with a non-default setting.
- Fixed a bug that could cause an invalid error message with regard to the database
  format in certain cases.
- The `--no-self-hits` option is no longer supported in `blastx` mode.
- Changed the semantics of the `--no-self-hits` option to check for equality of both
  sequence and sequence id, independent of the computed alignment.
- The selection of the top hit when using `--top` will respect the identity, coverage
  and no-self-hits filter settings (does not apply when frameshift alignment is
  enabled).
- The inclusion criterion for `--top` is applied to the bit score instead of the raw
  score and is no longer affected by integer rounding (does not apply when frameshift
  alignment is enabled).
- Improved the accuracy of the ranking heuristic.
- Added the options `--ext-chunk-size` and `--no-ranking` to control the ranking
  heuristic.

[2.0.2]
- Fixed a bug (persisting since v2.0.0) that could cause incomplete results in `blastx` mode.
- Reduced the use of temporary disk space.
- Fixed an issue that could cause long runtimes when using the `--taxon-list` option.

[2.0.1]
- Added feature for using the tool in a distributed computing environment.
- Fixed an issue that could cause increased memory usage and runtimes for very long queries.
- Fixed a bug that could cause a crash when using `--comp-based-stats 0`.
- Fixed a bug that could cause a crash for small input files in certain cases.
- Fixed a bug that could cause filtering hits for identity or range cover not to function correctly when using
  the tabular format without traceback being enabled.
- Added warning messages to recommend block size parameters based on system RAM.

[2.0.0]
- Added the sensitivity modes `--very-sensitive` and `--ultra-sensitive`.
- The `--block-size`/`-b` parameter is set to 0.4 and the `--index-chunks`/`-c` parameter is set to 1 by default
  in the new sensitivity modes.
- Improved performance.
- Added the option `--ext` with possible values `banded-fast` and `banded-slow` to adjust band setup for Smith
  Waterman extensions (new default is `banded-fast` for the default and sensitive mode, and `banded-slow` otherwise).
- Added automatic disabling of alignment traceback if not required by the user-defined output fields in tabular
  output format.
- Changed the default value of the `--max-hsps` parameter to 1.
- Changed the default value of the `--freq-sd` parameter from 10 to 20 for the sensitive mode.
- Fixed a compiler error on FreeBSD.

[0.9.36]
- Fixed a bug that could cause `makedb` to produce invalid database files when using taxonomy features.
- Fixed a bug that could cause a crash when running in query-indexed mode.

[0.9.35]
- Fixed a bug in `diamond view` that would cause high memory usage and erroneous output.
- Reduced the use of temporary disk space.
- Fixed a database compatibility issue with big endian architectures.
- Fixed a bug that would cause a crash for query sequences shorter than 5 letters in blastx mode.
- Fixed a bug that would cause a crash when using a FASTA file as database parameter in blastx mode.
- Added support for the following new ranks in the NCBI taxonomy: biotype, clade, forma specialis, genotype, isolate, morph, pathogroup, serogroup, serotype, strain, subvariety.

[0.9.34]
- Fixed a compiler error for native builds.
- Fixed a compiler error for GCC 4.8.
- Fixed a compiler error when support for SSSE3 was enabled without support for SSE4.1.
- Implemented asynchronous loading of seed hits.

[0.9.33]
- Improved performance and sensitivity.
- Increased use of temporary disk space.
- Implemented support for the AVX2 instruction set.
- Fixed a bug on big-endian architectures.
- Fixed bugs for compilers with unsigned char.
- Fixed compiler errors for generic builds.
- Added compatibility of database files between little and big endian architectures.
- Fixed various issues related to `Illegal instruction` errors on macOS.
- Added option `--file-buffer-size` to set the size of the I/O buffers and set the default value to 64 MB.

[0.9.32]
- Fixed a bug that would generate an incorrect count of positive scoring letters in all output formats.
- Fixed a compiler error on macOS.
- Fixed an `illegal instruction` error on macOS.

[0.9.31]
- Improved performance.
- Composition based statistics use integer scoring.
- Option `--quiet` will suppress startup message.
- Added output field `scovhsp` to print the subject coverage per HSP to the tabular format.

src/basic/config.h
/****
DIAMOND protein aligner
Copyright (C) 2013-2020 Max Planck Society for the Advancement of Science e.V.
                        Benjamin Buchfink
                        Eberhard Karls Universitaet Tuebingen

Code developed by Benjamin Buchfink <benjamin.buchfink@tue.mpg.de>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
****/

#pragma once
#include <string>
#include <vector>
#include <stdint.h>
#include <map>

enum class Sensitivity { FAST = 0, MID_SENSITIVE = 1, SENSITIVE = 2, MORE_SENSITIVE = 3, VERY_SENSITIVE = 4, ULTRA_SENSITIVE = 5 };
enum class TracebackMode { NONE = 0, SCORE_ONLY = 1, STAT = 2, VECTOR = 3, SCORE_BUFFER = 4 };

struct Config
{

	using string = std::string;
	using string_vector = std::vector<std::string>;

	string_vector input_ref_file;
	unsigned	threads_;
	string	database;
	string	query_file;
	unsigned	merge_seq_treshold;
	unsigned	hit_cap;
	unsigned shapes;
	unsigned	index_mode;
	size_t	max_alignments;
	string	match_file1;
	string	match_file2;
	int		padding;
	unsigned	output_threads;
	unsigned compression;
	unsigned		lowmem;
	double	chunk_size;
	unsigned min_identities;
	unsigned min_identities2;
	double ungapped_xdrop;
	int		raw_ungapped_xdrop;
	unsigned	min_compressed_identities;
	int		min_seed_score;
	unsigned	seed_signatures;
	double	min_bit_score;
	unsigned	run_len;
	double	max_seed_freq;
	string	tmpdir;
	string	parallel_tmpdir;
	bool		long_mode;
	int		gapped_xdrop;
	double	max_evalue;
	string	kegg_file;
	int		gap_open;
	int		gap_extend;
	string	matrix;
	bool debug_log, verbose, quiet;
	bool		salltitles;
	int		reward;
	int		penalty;
	string	db_type;
	double	min_id;
	unsigned	compress_temp;
	double	toppercent;
	string	daa_file;
	string_vector	output_format;
	string	output_file;
	bool		forwardonly;
	unsigned fetch_size;
	uint64_t	db_size;
	double	query_cover;

	bool		mode_sensitive;
	unsigned	verbosity;
	bool no_auto_append;
	unsigned local_align_mode;
	bool extend_all;
	bool slow_search;
	string_vector seq_no;
	double rank_factor;
	double rank_ratio;
	bool ht_mode;
	bool old_freq;
	double freq_sd;
	unsigned target_fetch_size;
	bool mode_more_sensitive;
	string matrix_file;
	double lambda, K;
	string_vector shape_mask;
	unsigned query_gencode;
	string unaligned;
	double space_penalty;
	bool new_prefilter;
	bool reverse;
	unsigned comp_based_stats;
	int neighborhood_score;
	unsigned seed_weight;
	int report_unaligned;
	double subject_cover;
	bool mode_very_sensitive;
	unsigned max_hsps;
	bool no_self_hits;
	unsigned id_left, id_right, id_n;
	int bmatch, bmismatch, bcutoff;
	unsigned query_bins;
	uint64_t n_ants;
	double rho;
	double p_best;
	double d_exp, d_new;
	double score_estimate_factor;
	int diag_min_estimate;
	string qfilt, sfilt;
	double path_cutoff;
	bool use_smith_waterman;
	string prot_accession2taxid;
	int superblock;
	unsigned max_cells;
	int masking;
	bool log_query;
	bool log_subject;
	unsigned threads_align;
	double score_ratio;
	bool small_query;
	bool hashed_seeds;
	string nodesdmp;
	bool sallseqid;
	string query_strands;
	bool xml_blord_format;
	int frame_shift;
	bool query_range_culling;
	double query_range_cover;
	double transcript_len_estimate;
	string family_counts_file;
	string taxonlist;
	bool radix_cluster_buffered;
	unsigned join_split_size;
	unsigned join_split_key_len;
	unsigned radix_bits;
	double join_ht_factor;
	bool sort_join;
	bool simple_freq;
	double freq_treshold;
	bool use_lazy_dict;
	string aligned_file;
	bool use_dataset_field;
	bool store_query_quality;
	string invocation;
	unsigned swipe_chunk_size;
	unsigned query_parallel_limit;
	bool long_reads;
	bool output_header;
	string alfmt;
	string unfmt;
	string namesdmp;
	bool hardmasked;
	int cbs_window;
	double tantan_r;
	double tantan_minMaskProb;
	bool no_unlink;
	bool no_dict;
	int stop_match_score;
	int tantan_maxRepeatOffset;
	bool tantan_ungapped;
	string taxon_exclude;
	bool swipe_all;
	uint64_t taxon_k;
	uint64_t upgma_edge_limit;
	string tree_file;
	string upgma_dist;
	string upgma_input;
	bool log_extend;
	int chaining_maxgap;
	string family_map;
	string family_map_query;
	size_t chaining_range_cover;
	bool no_swipe_realign;
	bool cut_bar;
	bool bootstrap;
	size_t chaining_maxnodes;
	int cutoff_score_8bit;
	double inner_culling_overlap;
	double min_band_overlap;
	int min_realign_overhang;
	int ungapped_window;
	int gapped_filter_diag_score;
	double gapped_filter_diag_bit_score;
	double gapped_filter_evalue;
	int gapped_filter_window;
	bool output_hits;
	double ungapped_evalue;
	bool no_logfile;
	bool no_heartbeat;
	int band_bin;
	int col_bin;
	size_t file_buffer_size;
	bool self;
	size_t trace_pt_fetch_size;
	uint32_t tile_size;
	double short_query_ungapped_bitscore;
	int short_query_max_len;
	double gapped_filter_evalue1;
	size_t ext_chunk_size;
	double ext_min_yield;
	string ext;
	int full_sw_len;
	double relaxed_evalue_factor;
	string type;
	bool raw;
	bool mode_ultra_sensitive;
	double chaining_len_cap;
	size_t chaining_min_nodes;
	bool fast_tsv;
	unsigned target_parallel_verbosity;
	double memory_limit;
	size_t global_ranking_targets;
	bool mode_mid_sensitive;
	bool no_ranking;
	bool query_memory;
	size_t memory_intervals;
	size_t seedhit_density;
	size_t chunk_size_multiplier;
	double ranking_score_drop_factor;
	double ranking_cutoff_bitscore;
	int left_most_interval;
	bool no_forward_fp;
	bool no_ref_masking;
	string roc_file;
	bool target_bias;
	bool check_multi_target;
	bool output_fp;
	int family_cap;
	int cbs_matrix_scale;
	size_t query_count;
	double cbs_angle;
	int target_seg;
	double cbs_err_tolerance;
	int cbs_it_limit;
	double query_match_distance_threshold;
	double length_ratio_threshold;

	Sensitivity sensitivity;
	TracebackMode traceback_mode;

	bool multiprocessing;
	bool mp_init;

	enum {
		makedb = 0, blastp = 1, blastx = 2, view = 3, help = 4, version = 5, getseq = 6, benchmark = 7, random_seqs = 8, compare = 9, sort = 10, roc = 11, db_stat = 12, model_sim = 13,
		match_file_stat = 14, model_seqs = 15, opt = 16, mask = 17, fastq2fasta = 18, dbinfo = 19, test_extra = 20, test_io = 21, db_annot_stats = 22, read_sim = 23, info = 24, seed_stat = 25,
		smith_waterman = 26, cluster = 27, translate = 28, filter_blasttab = 29, show_cbs = 30, simulate_seqs = 31, split = 32, upgma = 33, upgma_mc = 34, regression_test = 35,
		reverse_seqs = 36, compute_medoids = 37, mutate = 38, merge_tsv = 39, rocid = 40
	};
	unsigned	command;

	enum { double_indexed = 0, query_indexed = 1, subject_indexed = 2 };
	int algo;

	string cluster_algo;
	string_vector cluster_steps;
	double cluster_mcl_inflation;
	uint32_t cluster_mcl_expansion;
	double cluster_mcl_sparsity_switch;
	string	cluster_similarity;

	
	std::map<std::string, Sensitivity> sens_map{
		{"fast", Sensitivity::FAST},
		{"sensitive", Sensitivity::SENSITIVE},
		{"mid-sensitive", Sensitivity::MID_SENSITIVE},
		{"more-sensitive", Sensitivity::MORE_SENSITIVE},
		{"very-sensitive", Sensitivity::VERY_SENSITIVE},
		{"ultra-sensitive", Sensitivity::ULTRA_SENSITIVE}
	};

	enum { query_parallel = 0, target_parallel = 1 };
	unsigned load_balancing;

	Config() {}
	Config(int argc, const char **argv, bool check_io = true);

	inline unsigned get_run_len(unsigned length)
	{
		if (run_len == 0) {
			if (length < 30)
				return 1;
			else if (length < 100)
				return 20;
			else
				return 40;
		}
		else
			return run_len;
	}

	inline bool output_range(unsigned n_target_seq, int score, int top_score)
	{
		if (toppercent < 100)
			return (1.0 - (double)score / top_score) * 100 <= toppercent;
		else
			return n_target_seq < max_alignments;
	}

	void set_sens(Sensitivity sens);

	bool mem_buffered() const { return tmpdir == "/dev/shm"; }

  	template<typename _t>
	static void set_option(_t& option, _t value, _t def = 0) { if (option == def) option = value; }
};

void print_warnings();
extern Config config;

template<typename _t>
_t top_cutoff_score(_t top_score) {
	return _t((1.0 - config.toppercent / 100.0)*top_score);
}